### ISSUE TITLE
The following Cypher queries will trigger a database error on Neo4j 4.3.10

### ISSUE THREAD

**Issue by abc123000111** [neo4j/neo4j#12853]
Hi, I discovered that the following Cypher queries will trigger a database error (Neo.DatabaseError.Statement.ExecutionFailed). I wonder that if it is a bug. Thank you!

**Neo4j Version:** 4.3.10 Community
**Operating System:** Windows 11
**API:** Cypher / Java API

### Steps to reproduce
Executing the following queries:

```
CREATE (n0 :L1{k7:-707619519}) RETURN *;
CREATE (n1 :L0{k3:true}) RETURN *;
CREATE (n2 :L0 :L2{k2:-122810581}) RETURN *;
CREATE (n3) RETURN *;
CREATE (n4 :L1{k6:true, k8:'hy2Fa3O'}) RETURN *;
CREATE (n5) RETURN *;
CREATE (n6) RETURN *;
CREATE (n7) RETURN *;
CREATE (n8) RETURN *;
CREATE (n9 :L0{k0:-1215419690}) RETURN *;
CREATE (n10) RETURN *;
CREATE (n11 :L2{k14:'eG3'}) RETURN *;
CREATE (n12) RETURN *;
CREATE (n13) RETURN *;
MATCH (n4 :L1{k6:true, k8:'hy2Fa3O'}), (n13) MERGE (n4)-[r0 :T1{k25:'6'}]->(n13) RETURN *;
MATCH (n0) MATCH (n1)--(n2) OPTIONAL MATCH (n1), (n0)-[r2 *1..]-(n3:L1) RETURN 0;
```

### Expected behavior
No error occurred.

### Actual behavior
**Neo.DatabaseError.Statement.ExecutionFailed**
Expected 
RegularSinglePlannerQuery(QueryGraph {Nodes: ['n0', 'n1', 'n2'], Rels: ['(n1)-[  UNNAMED0]-(n2)'], Optional Matches: : ['QueryGraph {Nodes: ['n0', 'n1', 'n3'], Rels: ['(n0)-[r2*]-(n3)'], Arguments: ['n0', 'n1'], Predicates: ['n3:L1']}']},InterestingOrder(RequiredOrderCandidate(List()),List()),RegularQueryProjection(Map(0 -> AutoExtractedParameter(  AUTOINT0,Integer,SignedDecimalIntegerLiteral(0))),QueryPagination(None,None),Selections(Set())),None,None) 

Instead, got: 
RegularSinglePlannerQuery(QueryGraph {Nodes: ['n0', 'n1', 'n2'], Rels: ['(n1)-[  UNNAMED0]-(n2)'], Optional Matches: : ['QueryGraph {Nodes: ['n0', 'n3'], Rels: ['(n0)-[r2*]-(n3)'], Arguments: ['n0', 'n1'], Predicates: ['n3:L1']}']},InterestingOrder(RequiredOrderCandidate(List()),List()),RegularQueryProjection(Map(0 -> AutoExtractedParameter(  AUTOINT0,Integer,SignedDecimalIntegerLiteral(0))),QueryPagination(None,None),Selections(Set())),None,None)
Plan: ProduceResult(List(0)) {
  LHS -> Projection(Map(0 -> AutoExtractedParameter(  AUTOINT0,Integer,SignedDecimalIntegerLiteral(0)))) {
    LHS -> Apply(false) {
      LHS -> CartesianProduct(false) {
        LHS -> Expand(n1, BOTH, List(), n2,   UNNAMED0, ExpandAll) {
          LHS -> AllNodesScan(n1, Set()) {}
        }
        RHS -> AllNodesScan(n0, Set()) {}
      }
      RHS -> Optional(Set(n1,   UNNAMED0, n2, n0)) {
        LHS -> VarExpand(n0, BOTH, OUTGOING, List(), n3, r2, VarPatternLength(1,None), ExpandInto, None, None) {
          LHS -> NodeByLabelScan(n3, LabelName(L1), Set(n0, n1), IndexOrderNone) {}
        }
      }
    }
  }
} 

Differences:
 - QueryGraph
    A: QueryGraph {Nodes: ['n0', 'n1', 'n2'], Rels: ['(n1)-[  UNNAMED0]-(n2)'], Optional Matches: : ['QueryGraph {Nodes: ['n0', 'n1', 'n3'], Rels: ['(n0)-[r2*]-(n3)'], Arguments: ['n0', 'n1'], Predicates: ['n3:L1']}']}
    B: QueryGraph {Nodes: ['n0', 'n1', 'n2'], Rels: ['(n1)-[  UNNAMED0]-(n2)'], Optional Matches: : ['QueryGraph {Nodes: ['n0', 'n3'], Rels: ['(n0)-[r2*]-(n3)'], Arguments: ['n0', 'n1'], Predicates: ['n3:L1']}']}

**Comment by HannesSandberg** [comment]
Thank you! We will look into if this is a bug and come back to you.

**Comment by arnefischereit** [comment]
Hi @abc123000111,

This is caused by the same problem as #12818 and a fix will be released with the upcoming 4.3.12 release.
